oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/MongoDownloaderRegexUtils.java

```diff
@@ -0,0 +1,170 @@
+        }
+
+        if (filters.isEmpty()) {
+            return null;
+        } else if (filters.size() == 1) {
+            return filters.get(0);
+        } else {
+            return Filters.and(filters);
+        }
+    }
+
+    private static List<Pattern> compileIncludedDirectoriesRegex(List<String> includedPaths) {
+        return compileDirectoryRegex(includedPaths, false);
+    }
+
+    private static List<Pattern> compileExcludedDirectoriesRegex(List<String> excludedPaths) {
+        return compileDirectoryRegex(excludedPaths, true);
+    }
+
+    private static List<Pattern> compileDirectoryRegex(List<String> paths, boolean negate) {
+        if (paths.isEmpty()) {
+            return List.of();
+        }
+        if (paths.size() == 1 && paths.get(0).equals("/")) {
+            return List.of();
+        }
+        ArrayList<Pattern> patterns = new ArrayList<>();
+        for (String path : paths) {
+            if (!path.endsWith("/")) {
+                path = path + "/";
+            }
+            String regex = "^[0-9]{1,3}:" + Pattern.quote(path);
+            Pattern pattern;
+            if (negate) {
+                pattern = compileExcludedDirectoryRegex(regex);
+            } else {
+                pattern = Pattern.compile(regex);
+            }
+            patterns.add(pattern);
+        }
+        return patterns;
+    }
+
+    static Pattern compileExcludedDirectoryRegex(String regex) {
+        // https://stackoverflow.com/questions/1240275/how-to-negate-specific-word-in-regex
+        return Pattern.compile("^(?!" + regex + ")");
+    }
+
+    static Pattern compileCustomExcludedPatterns(String customRegexPattern) {
+        if (customRegexPattern == null || customRegexPattern.trim().isEmpty()) {
+            LOG.info("Mongo custom regex is disabled");
+            return null;
+        } else {
+            LOG.info("Excluding nodes with paths matching regex: {}", customRegexPattern);
+            var negatedRegex = "^(?!.*(" + customRegexPattern + ")$)";
+            return Pattern.compile(negatedRegex);
+        }
+    }
+
+    /**
+     * Returns all the ancestors paths of the given list of paths. That is, if the list is ["/a/b/c", "/a/b/d"],
+     * this method will return ["/", "/a", "/a/b", "/a/b/c", "/a/b/d"]. Note that the paths on the input list are also
+     * returned, even though they are not strictly ancestors of themselves.
+     */
+    static List<String> getAncestors(List<String> paths) {
+        TreeSet<String> ancestors = new TreeSet<>();
+        for (String child : paths) {
+            String parent = child;
+            while (true) {
+                ancestors.add(parent);
+                if (PathUtils.denotesRoot(parent)) {
+                    break;
+                }
+                parent = PathUtils.getParentPath(parent);
+            }
+        }
+        return new ArrayList<>(ancestors);
+    }
```

Comment:
Since we only use the TreeSet to maintain a sorted order in the end (and not intermittently) I think we could use a hash set and sort in the end, but that's just a minor thing. In theory it's still the same O-complexity, but in practice I think it should be quite faster, as because of the cache locality benefits of sorting an `ArrayList`. 

```suggestion
static List<String> getAncestors(List<String> paths) {
    HashSet<String> ancestors = new HashSet<>();
    for (String child : paths) {
        String parent = child;
        while (!PathUtils.denotesRoot(parent)) {
            ancestors.add(parent);
            parent = PathUtils.getParentPath(parent);
        }
        ancestors.add(parent); // add the root path as well
    }
    List<String> sortedAncestors = new ArrayList<>(ancestors);
    Collections.sort(sortedAncestors);
    
    return sortedAncestors;
}
```
But I would like change the `while(true)` to condition on the the root check instead.

Reply:
This code is not on the critical path, so I prefer to use the TreeSet to both sort and remove duplicates. I have simplified the while loop as suggested.